sycl/include/sycl/detail/vector_arith.hpp

```diff
@@ -24,49 +24,166 @@ template <typename DataT, int NumElem> class __SYCL_EBO vec;
+    return std::forward<T>(lhs) >> std::forward<U>(rhs);
+  }
+};
 
-// Macros to populate binary operation on sycl::vec.
-#if defined(__SYCL_BINOP) || defined(BINOP_BASE)
-#error "Undefine __SYCL_BINOP and BINOP_BASE macro"
-#endif
+struct UnaryPlus {
+  template <class T>
+  constexpr auto operator()(T &&arg) const -> decltype(+std::forward<T>(arg)) {
+    return +std::forward<T>(arg);
+  }
+};
 
+struct VecOperators {
 #ifdef __SYCL_DEVICE_ONLY__
-#define BINOP_BASE(BINOP, OPASSIGN, CONVERT, COND)                             \
-  template <typename T = DataT>                                                \
-  friend std::enable_if_t<(COND), vec_t> operator BINOP(const vec_t & Lhs,     \
-                                                        const vec_t & Rhs) {   \
-    vec_t Ret;                                                                 \
-    if constexpr (vec_t::IsBfloat16) {                                         \
-      for (size_t I = 0; I < NumElements; ++I) {                               \
-        Ret[I] = Lhs[I] BINOP Rhs[I];                                          \
-      }                                                                        \
-    } else {                                                                   \
-      auto ExtVecLhs = sycl::bit_cast<typename vec_t::vector_t>(Lhs);          \
-      auto ExtVecRhs = sycl::bit_cast<typename vec_t::vector_t>(Rhs);          \
-      Ret = vec<DataT, NumElements>(ExtVecLhs BINOP ExtVecRhs);                \
-      if constexpr (std::is_same_v<DataT, bool> && CONVERT) {                  \
-        vec_arith_common<bool, NumElements>::ConvertToDataT(Ret);              \
-      }                                                                        \
-    }                                                                          \
-    return Ret;                                                                \
+  static constexpr bool is_host = false;
+#else
+  static constexpr bool is_host = true;
+#endif
+
+  template <typename BinOp, typename Self>
+  static constexpr auto apply_binary_op(const Self &Lhs, const Self &Rhs) {
+    static_assert(is_vec_v<Self>);
+    using element_type = typename Self::element_type;
+    constexpr int N = Self::size();
+    constexpr bool is_logical =
+        check_type_in_v<BinOp, std::equal_to<void>, std::not_equal_to<void>,
+                        std::less<void>, std::greater<void>,
+                        std::less_equal<void>, std::greater_equal<void>,
+                        std::logical_and<void>, std::logical_or<void>>;
+    using result_t = std::conditional_t<
+        is_logical, vec<fixed_width_signed<sizeof(element_type)>, N>, Self>;
+
+    BinOp Op{};
+    if constexpr (is_host ||
+                  std::is_same_v<element_type, ext::oneapi::bfloat16>) {
+      result_t res{};
+      for (size_t i = 0; i < N; ++i)
+        if constexpr (is_logical)
```

Comment:
could you remind me why we are using -1 on the logical ops here? Or maybe insert a comment?

Reply:
Because that's how it's defined in the SYCL spec... ¯\_(ツ)_/¯

I'm doing some minor changes here to unify several code paths and fix some CTS failures.